Port of point_cloud2.py from ROS1 to ROS2.

SebastianGrans opened the pull request:
Hi, 

The python submodule `point_cloud2` in ROS1 was useful when working with `PointCloud2` messages. 
[People](https://answers.ros.org/question/357944/why-is-point_cloud2py-missing-from-sensor_msgs-in-ros2/) including me would like this feature in ROS2, so I took the opportunity to "port" it.

This is the first time I ever perform a PR to a significant project, and I've tried to follow the developer guidelines to the best of my abilities. Neverthessless, I've probably made some mistakes. 

**Tests:**  
Not all tests are passing. 

```
The following tests FAILED:
	 27 - flake8_rosidl_generated_py (Failed)
	 34 - flake8 (Failed)
	 36 - copyright (Failed)
```
The two former seem to not be related to this package per se, but rather to `flake8` [link](https://gitlab.com/pycqa/flake8/-/issues/640). Here's the error:
```
27: Traceback (most recent call last):      
27:   File "/opt/ros/eloquent/bin/ament_flake8", line 11, in <module>
27:     load_entry_point('ament-flake8==0.8.1', 'console_scripts', 'ament_flake8')()
27:   File "/opt/ros/eloquent/lib/python3.6/site-packages/ament_flake8/main.py", line 75, in main
27:     max_line_length=args.linelength)
27:   File "/opt/ros/eloquent/lib/python3.6/site-packages/ament_flake8/main.py", line 163, in generate_flake8_report
27:     style = get_flake8_style_guide(flake8_argv)
27:   File "/opt/ros/eloquent/lib/python3.6/site-packages/ament_flake8/main.py", line 131, in get_flake8_style_guide
27:     application.parse_preliminary_options_and_args([])
27: AttributeError: 'Application' object has no attribute 'parse_preliminary_options_and_args'
```

And the `copyright` test gives me:
```
36: sensor_msgs/point_cloud2.py: copyright=Willow Garage, Inc. (2008), license=<unknown>
36: test/test_point_cloud2.py: copyright=Sebastian Grans (2020), license=<unknown>
```
`point_cloud2.py` has the original copyright header from [here](https://github.com/ros/common_msgs/blob/noetic-devel/sensor_msgs/src/sensor_msgs/point_cloud2.py). ~~And for `test/test_point_cloud2.py` I simply copied the Apache License header from `test_pointcloud_conversion.cpp`.~~ (Apparently not.)

Signed-off-by: Sebastian Grans <sebastian.grans@gmail.com>

clalancette commented:
I'm not really a big fan of making mixed C++/python packages.  It tends to make things not work that great for either, and it is hard to install one without installing the other.

There are at least 2 ways we can handle this:
1.  Make a separate package within common_interfaces that has just this file in it (`point_cloud_py` or something like that).
1.  Make a completely separate repository with this code in it.  I don't particularly find this code to be "core" to ROS 2, so this makes some sense to me.

I'd personally go with the second option, since I don't think we need to host this code in the core here.  But I'd like to hear opinions from @ros2/team .

SebastianGrans commented:
I think I agree with the second option, but let's wait and see what the others think.

dirk-thomas commented:
> It tends to make things not work that great for either

As long as you don't want to add an entry point to the Python part I am not aware of any limitation in mixing Python modules with a CMake package.

wjwwood commented:
I think a big drawback is making it harder to separate Python and C++ parts of our code base, e.g. if you only want to install C++ stuff and actually don't want a runtime dependency on Python or Python libraries. As you go up in the stack that become less important, but in a core package like `common_interfaces`, it's more impactful.

That all being said, I'm fine with leaving it as-is, but if there's motivation or another reason to separate the Python code out, then I'd agree with @jacobperron in that I have no preference as to whether it is in this repository or not.

dmorris0 commented:
I think this functionality is needed by just about anyone who uses ROS2-Python to subscribe to and process point clouds -- which I see as a big use for ROS2.  So I ideally it will be part of ROS2.

clalancette commented:
> As long as you don't want to add an entry point to the Python part I am not aware of any limitation in mixing Python modules with a CMake package.

Yeah, that's the downside I was thinking about.  I guess it doesn't apply here, but I don't see a reason to prevent that, and it just seems cleaner to have separate packages for Python and C++.

> I think a big drawback is making it harder to separate Python and C++ parts of our code base, e.g. if you only want to install C++ stuff and actually don't want a runtime dependency on Python or Python libraries. As you go up in the stack that become less important, but in a core package like `common_interfaces`, it's more impactful.

Exactly for this reason.

> I think this functionality is needed by just about anyone who uses ROS2-Python to subscribe to and process point clouds -- which I see as a big use for ROS2. So I ideally it will be part of ROS2.

While I agree that a lot of ROS users use PCL things, I still don't think it should be in the core like this.  Most of the C++ code to deal with PCL in C++ lives in https://github.com/ros-perception/perception_pcl repository, for instance.  I think this fits in better there, so my suggestion is to create a new Python package over there and port this functionality there.  It may be worthwhile to open an issue there first to make sure the maintainers agree with that plan.

mikaelarguedas commented:
+1 it would be great to have these utilities available in ROS 2.

In general it makes sense to have the utilities to operate on a data structure close to the definition of the data structure itself so I'd find it sensible to have this in this package or this repo (akin [similar C++ utilities](https://github.com/ros2/common_interfaces/blob/c1c41310c4373a197dcbd1bd65bafc34a55e1255/sensor_msgs/include/sensor_msgs/point_cloud2_iterator.hpp) that live within sensor_msgs).

IIUC perception_pcl packages serve a different purpose which is to easily connect the ROS PointCloud2 datastructure to the PCL back-end for processing. These utilities being independent of PCL it would break the abstraction between the ROS datastructure and the library used for processing, It could be confusing to have the utilities live there and imply to some extent that the PointCloud2 datastructure and its utilities cannot be used with other processing approaches (being custom code or other point cloud libraries like open3d).

mabelzhang commented:
Trying to push this forward.

I see that two things are being discussed:
1. Where should this code go, core stacks or `perception_pcl`
2. If the answer is core stacks, should it go into this C++ package or a separate Python package

As for number 1, as mikaelarguedas pointed out and linked to, C++ utilities that iterate through `PointCloud2` and provide reading and writing are currently being hosted in `common_interfaces`. The Python utility being added in this PR has similar functionalities - creating and reading `PointCloud2`. So it makes sense for at least these two utilities to be in the same place. If we move the Python script to `perception_pcl`, then we should probably move the C++ utilities as well so that people can find both in the same place, and I don't think we want to move the existing C++ stuff. So, I gather the answer to number 1 is that it should go to core stacks.

Regarding number 2, I count 1 vote for moving this to a separate Python package, 3 votes indifferent. As wjwwood and clalancette pointed out, keeping it here has a drawback, separating it out has no drawback. So in the long run, it seems more optimal to do the way without a drawback. That does not involve too much change in this PR, just creating a new package in this repo.

Looking at the list of packages in this repo, I don't know what package makes sense though, as everything in this repo are message packages, and `sensor_msgs` actually seems the reasonable place. Doing a `sensor_msgs_py` seems it'd make a predecessor for a bunch of Python packages, one for each type of messages. Maybe we could have a new package just for Python, with a name generic enough to encompass all messages in this repo, and future Python scripts can be dumped there.
@clalancette thoughts, since you voted for separating Python from C++?

jacobperron commented:
:+1: for `sensor_msgs_py`.

mabelzhang commented:
Looks like `sensor_msgs_py` it is.

@SebastianGrans Per discussion above, could you please refactor the code such that it resides in a new Python-only package named `sensor_msgs_py` in this repo?

SebastianGrans commented:
Hm... This is embarrassing. I reverted my fork and refactored the code to a separate package. It seems like that is not the way you are supposed to do it, since it auto-closed this PR. 

What should I do now? Do I make have to make a new PR?

I'm sorry for any inconvenience.

clalancette commented:
It looks like you got your commits sorted, so that is good.

You are currently failing both of the automated checks.

The first one is the DCO bot, which you have to sign-off *all* of your commits to pass.  I'll suggest that you rebase and squash all of the current commits together, and just sign-off the one commit.

The second one is failing to complete the tests for the new package.  The error message is:

```
ImportError while importing test module '/tmp/ws/src/common_interfaces/sensor_msgs_py/test/test_point_cloud2.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test/test_point_cloud2.py:7: in <module>
    from sensor_msgs.msg import PointCloud2
E   ModuleNotFoundError: No module named 'sensor_msgs'
```

This looks like you need to update the tests for the new name of the package.  There are also Python warnings in there; I'll suggest you run the tests locally and resolve all of those.

SebastianGrans commented:
That required a bit of learning. Feels great to get to know Git a bit better though. 

I think everything should be in order. Though I still get an error from the copyright test.
```
colcon test-result --verbose                
build/sensor_msgs_py/pytest.xml: 9 tests, 0 errors, 1 failure, 0 skipped
- sensor_msgs_py.test.test_copyright test_copyright (test/test_copyright.py:18)
  <<< failure message
    AssertionError: Found errors assert 1 == 0
  >>>

Summary: 9 tests, 0 errors, 1 failure, 0 skipped
``

clalancette commented:
> I think everything should be in order. Though I still get an error from the copyright test.

Hm, yeah, that's unfortunate.  I think the problem is that our copyright checker doesn't support BSD properly at the moment.  I'll suggest just removing that test for now.

Also, the DCO check is still failing.  I believe that is because the email address you have configured in your `.gitconfig` file is different than the one you are using to sign-off with.  If you could reconcile those, the bot will be more happy.

clalancette commented on sensor_msgs_py/package.xml:
```diff
@@ -0,0 +1,32 @@
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>sensor_msgs_py</name>
+  <version>1.0.0</version>
```
This needs to be the same version as the other packages already in this repository.  In this case, it looks like 2.1.0.

clalancette commented on sensor_msgs_py/package.xml:
```diff
@@ -0,0 +1,32 @@
+  <version>1.0.0</version>
+  <description>A package for easy creation and reading of PointCloud2 messages in Python.</description>
+  <maintainer email="sebastian.grans@gmail.com">Sebastian Grans</maintainer>
+  <license>Apache License 2.0</license>
```
Based on the rest of the code, this should be BSD.

clalancette commented on sensor_msgs_py/package.xml:
```diff
@@ -0,0 +1,32 @@
+  <maintainer email="sebastian.grans@gmail.com">Sebastian Grans</maintainer>
+  <license>Apache License 2.0</license>
+
+  <buildtool_depend>rosidl_default_generators</buildtool_depend>
```
This dependency shouldn't be necessary in this package.

clalancette commented on sensor_msgs_py/package.xml:
```diff
@@ -0,0 +1,32 @@
+
+  <buildtool_depend>rosidl_default_generators</buildtool_depend>
+
+  <build_depend>builtin_interfaces</build_depend>
+  <build_depend>geometry_msgs</build_depend>
+  <build_depend>std_msgs</build_depend>
```
Python packages don't typically have any `build_depend`, so you can remove all of these.

clalancette commented on sensor_msgs_py/package.xml:
```diff
@@ -0,0 +1,32 @@
+  <build_depend>geometry_msgs</build_depend>
+  <build_depend>std_msgs</build_depend>
+
+  <exec_depend>builtin_interfaces</exec_depend>
+  <exec_depend>geometry_msgs</exec_depend>
+  <exec_depend>rosidl_default_runtime</exec_depend>
```
These three shouldn't be necessary; this package doesn't directly depend on any of them.

clalancette commented on sensor_msgs_py/sensor_msgs_py/point_cloud2.py:
```diff
@@ -0,0 +1,232 @@
+"""
+
+
+from collections import namedtuple
```
This is throwing a warning in the CI job:

```

/usr/lib/python3/dist-packages/pydocstyle/config.py:6
  /usr/lib/python3/dist-packages/pydocstyle/config.py:6: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.9 it will stop working
    from collections import Set, namedtuple
```

I *think* you can fix it with `from collections.abc import namedtuple`, but I'm not sure.

SebastianGrans replied:
I can't fix this one. 

The error comes from the pydocstyle package. 
The default (apt) version of this package on Ubuntu is 2.1.1, but via pip you can get 5.1.1 where this if fixed.

clalancette replied:
Ah, OK.  Thanks for looking into it.  We'll just leave it like it is for now.

SebastianGrans commented:
Thanks! 
I'll get to fixit that tomorrow :)

clalancette commented on sensor_msgs_py/pytest.ini:
```diff
@@ -0,0 +1,2 @@
+[pytest]
+junit_family=xunit2
```
Nit: add a trailing newline.

clalancette commented on sensor_msgs_py/setup.py:
```diff
@@ -0,0 +1,25 @@
+
+setup(
+    name=package_name,
+    version='1.0.0',
```
This version should match what is in the package.xml, so change this to 2.0.1.

clalancette commented on sensor_msgs_py/setup.py:
```diff
@@ -0,0 +1,25 @@
+    description='A package for easy creation and reading of PointCloud2 messages in Python.',
+    license='Apache License 2.0',
+    tests_require=['pytest'],
+    entry_points={
+        'console_scripts': [
+        ],
+    },
```
Since this is empty, just remove this code completely.

clalancette commented on sensor_msgs_py/setup.py:
```diff
@@ -0,0 +1,25 @@
+    maintainer='Sebastian Grans',
+    maintainer_email='sebastian.grans@gmail.com',
+    description='A package for easy creation and reading of PointCloud2 messages in Python.',
+    license='Apache License 2.0',
```
This should match what is in the package.xml, so change this to 'BSD'.

clalancette commented on sensor_msgs_py/test/test_flake8.py:
```diff
@@ -0,0 +1,25 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
Change this to the BSD license to match the declared license of the package.

SebastianGrans replied:
In `point_cloud2.py` the the copyright is given as `Copyright (c) 2008, Willow Garage, Inc.`
Should this be used here as well? Both? 

And what about the year, should that be updated?

clalancette replied:
> In `point_cloud2.py` the the copyright is given as `Copyright (c) 2008, Willow Garage, Inc.`
> Should this be used here as well? Both?

This should be changed to be `Copyright (c) 2020 Open Source Robotics Foundation, Inc.`.  The reasoning here is that this is derived from other OSRF code, so it makes sense to have this copyright be OSRF as well.

> And what about the year, should that be updated?

Yeah, the year should be updated to 2020.

clalancette commented on sensor_msgs_py/test/test_pep257.py:
```diff
@@ -0,0 +1,23 @@
+# Copyright 2015 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
Same; change this to the BSD license.

SebastianGrans replied:
Same question as for `test_flake8.py`

clalancette commented on sensor_msgs_py/test/test_point_cloud2.py:
```diff
@@ -0,0 +1,111 @@
+
+
```
Please put the BSD license at the top here.

SebastianGrans replied:
Since I wrote this code, am I given the copyright? Or do I, by contributing, give the copyright to OSRF?

clalancette replied:
> Since I wrote this code, am I given the copyright? Or do I, by contributing, give the copyright to OSRF?

You maintain copyright to it, since you wrote it.

mabelzhang commented on sensor_msgs_py/test/test_point_cloud2.py:
```diff
@@ -0,0 +1,111 @@
+from sensor_msgs.msg import PointCloud2
+from sensor_msgs.msg import PointField
+from sensor_msgs_py import point_cloud2
+from std_msgs.msg import Header
```
`std_msgs` might need to be added to one of the depend tags in `package.xml`

mabelzhang commented on sensor_msgs_py/test/test_point_cloud2.py:
```diff
@@ -0,0 +1,111 @@
+        self.assertTrue(np.allclose(pcd_list, points_nonan, equal_nan=False))
+
+    def test_read_points_list(self):
+        # Suggestions on a reasonable test case?
```
What is this function intended to test for? Converting a list to a PointCloud2?

mabelzhang reviewed (commented):
Is the empty text file `resource/sensor_msgs_py` necessary? Or is there supposed to be something in it? I see it referenced in `setup.py` but I'm not sure what it is being used for.

clalancette commented:
> Is the empty text file `resource/sensor_msgs_py` necessary? Or is there supposed to be something in it? I see it referenced in `setup.py` but I'm not sure what it is being used for.

That empty file is needed in all pure Python packages.  I'm honestly not sure why, but I know it is needed.

sloretz commented:
> >    Is the empty text file resource/sensor_msgs_py necessary? Or is there supposed to be something in it? I see it referenced in setup.py but I'm not sure what it is being used for.
> 
> That empty file is needed in all pure Python packages. I'm honestly not sure why, but I know it is needed.

It's a file that get's copied into the ament_index. IIUC the Python packages have this as a file in the repo because `setup.py` can install files by copying them ([example](https://github.com/ros2/ros2cli/blob/84ce52f3a8327c1fc22d0297652a34ad3d0b58ec/ros2pkg/setup.py#L12-L13)), but it can't outright create one. See this documentation for more info about why the file is installed at all: https://github.com/ament/ament_cmake/blob/master/ament_cmake_core/doc/resource_index.md#file-system-index-layout

SebastianGrans commented:
Sorry for the late update.
I went on vacation and completely forgot about this PR.

clalancette commented on sensor_msgs_py/sensor_msgs_py/point_cloud2.py:
```diff
@@ -0,0 +1,231 @@
+#!/usr/bin/env python
+
```
Remove these two lines; this only works on Unix.

clalancette commented on sensor_msgs_py/sensor_msgs_py/point_cloud2.py:
```diff
@@ -0,0 +1,231 @@
+Serialization of sensor_msgs.PointCloud2 messages.
+
+Author: Tim Field
+ROS2 port by Sebastian Grans
```
```suggestion
ROS 2 port by Sebastian Grans
```

clalancette replied:
Also, can you please put a comment with a permalink saying where this originally was ported from?  Something like:

```
File originally ported from https://github.com/ros/common_msgs/blob/f48b00d43cdb82ed9367e0956db332484f676598/sensor_msgs/src/sensor_msgs/point_cloud2.py
```

clalancette reviewed (changes requested):
A few more minor changes, then we are good to run CI.

rafaelRinconFo commented:
@SebastianGrans Hi Sebastian. I would like to ask you something. The `read_points` function returns a giant list, however, I can't see the discriminated values of each field xyz. How is this list supposed to be read? How can I extract the data from this list?

flynneva commented:
@SebastianGrans I tried to help out and I fixed the left-over ament errors (flake8 and copyright) and [opened a PR](https://github.com/SebastianGrans/common_interfaces/pull/1) on your repo that you should be able to just review and merge into your branch

SebastianGrans commented:
@flynneva: Thanks! 

Sorry for the delay everyone.

clalancette commented on sensor_msgs_py/package.xml:
```diff
@@ -0,0 +1,19 @@
+  <name>sensor_msgs_py</name>
+  <version>2.1.0</version>
+  <description>A package for easy creation and reading of PointCloud2 messages in Python.</description>
+  <maintainer email="sebastian.grans@gmail.com">Sebastian Grans</maintainer>
+  <license>BSD</license>
```
It mostly only makes sense to have the maintainers of the repository be the maintainers of the sub-packages.  Thus, I suggest you change this to the following:

```suggestion
  <maintainer email="michael@openrobotics.org">Michael Carroll</maintainer>
  <maintainer email="michel@ekumenlabs.com">Michel Hidalgo</maintainer>
  <license>BSD</license>
  <author email="sebastian.grans@gmail.com">Sebastian Grans</author>
```

SebastianGrans replied:
Done! :)

clalancette commented on sensor_msgs_py/sensor_msgs_py/point_cloud2.py:
```diff
@@ -0,0 +1,231 @@
+# Copyright 2020 Open Source Robotics Foundation, Inc. (2020)
+# All rights reserved.
+#
+# Software License Agreement (BSD License 2.0)
```
This is totally not your fault, but we should fix up this license header.  This version of the header is a kind of mixed-up BSD 3-clause license.  The "official" license header is here: https://github.com/ament/ament_lint/blob/master/ament_copyright/ament_copyright/template/bsd_3clause_header.txt

(the same goes for all of the headers below)

The other problem we now have in this repository is that there is a top-level LICENSE file that specifies Apache 2.0, but that is no longer true for the entire repository.  We could do one of two things here:

1.  Rename the top-level LICENSE file to LICENSE.apache2 , and then add in LICENSE.bsd3 alongside it.
1.  Copy the current Apache 2 LICENSE file to each of the existing packages, then add in a new LICENSE file to this package with the BSD license in it.

I'm kind of leaning more towards the second option, but I'd like to hear any additional opinions from @ros2/team

flynneva replied:
@clalancette wow that license header thing is frustrating. i didnt know there was two different bsd 3 clause licenses in ament_copyright. why is there two different of the same license? and why is one "slightly" more right?

clalancette replied:
> @clalancette wow that license header thing is frustrating. i didnt know there was two different bsd 3 clause licenses in ament_copyright. why is there two different of the same license? and why is one "slightly" more right?

When I originally added it, I made a mistake and put slightly wrong text in there.  We've been slowly deprecating and removing that slightly incorrect version, but we haven't totally gotten rid of it yet.  So it is still there for legacy reasons; one of these days we'll remove it.  In any case, we shouldn't add new code that depends on it.

flynneva replied:
@clalancette why not just change this package to Apache 2.0 as well to match the rest of the packages in this repo? seems kind of arbitrary to change only this package to BSD 2.0.

clalancette replied:
> @clalancette why not just change this package to Apache 2.0 as well to match the rest of the packages in this repo? seems kind of arbitrary to change only this package to BSD 2.0.

Since this is a port of existing functionality, we can't unilaterally do that unless we have agreement from all of the previous authors.

flynneva replied:
@clalancette it looks like from [the original file](https://github.com/ros/common_msgs/blame/noetic-devel/sensor_msgs/src/sensor_msgs/point_cloud2.py) that @tfoote  or @dirk-thomas  could sign-off and say if we can switch to Apache 2.0 like the rest of this repo.

could either @tfoote or @dirk-thomas  sign off and agree that this package could be Apache 2.0 now to match the parent repo?

flynneva replied:
@sebastiangrans ive made another [PR to your branch](https://github.com/SebastianGrans/common_interfaces/pull/2) that switches this repo to Apache 2.0 to match the parent repo.

flynneva replied:
@clalancette should the copyright name then still be kept as the original `Willow Garage, Inc.`?

clalancette replied:
Oh yeah, good point.  It should be kept as the original then.

flynneva replied:
@clalancette @sebastiangrans PR [open here](https://github.com/SebastianGrans/common_interfaces/pull/4) with the discussed changes. i think i got them all...but definitely still check it

clalancette replied:
That looks reasonable to me.  As long as the linters pass, I'm happy with it.  We just need @SebastianGrans to merge that into this PR, then we can move forward here.

SebastianGrans replied:
Merged! :)

clalancette commented on sensor_msgs_py/setup.py:
```diff
@@ -0,0 +1,21 @@
+    ],
+    install_requires=['setuptools'],
+    zip_safe=True,
+    maintainer='Sebastian Grans',
+    maintainer_email='sebastian.grans@gmail.com',
```
Same thing about the maintainers here; I'll suggest making the maintainers the same as the package.

flynneva replied:
@christophebedard i didnt do this change [on my PR](https://github.com/SebastianGrans/common_interfaces/pull/4) because i figured we should probably get @wjwwood OK first.

SebastianGrans replied:
As far as I understand, `setup.py` only supports one maintainer. So I added Michael Carroll as the maintainer.

clalancette reviewed (changes requested):
I found a few more things that we need to clean up here.  Thanks for the iterations so far.

clalancette commented:
All right, I did some additional changes here to cleanup the copyrights, add in copyright checking, remove some backwards compatibility code, and fix some flake8 errors.  I'm now pretty happy with this, so I'll approve and run CI.  However, I would like to have someone else from the @ros2/team take a second look since I did some edits here.

clalancette commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=13226)](http://ci.ros2.org/job/ci_linux/13226/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=8156)](http://ci.ros2.org/job/ci_linux-aarch64/8156/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10941)](http://ci.ros2.org/job/ci_osx/10941/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=13256)](http://ci.ros2.org/job/ci_windows/13256/)

flynneva commented:
> remove some backwards compatibility code
@clalancette just for my understanding is there a main reason to do this? i use that backwards compatibility `main_with_errors` fix for my repos and wouldnt mind learning why **not** to do this.

clalancette commented:
> @clalancette just for my understanding is there a main reason to do this?

The main reason is that on master/Rolling, it is dead code.  It will never be executed there, so there is no reason to have it.  If we end up backporting this to distributions that don't have `main_with_errors` (which I guess at this point would only be Dashing), then we would reintroduce the compatibility code (but only on the `dashing` branch).

> i use that backwards compatibility `main_with_errors` fix for my repos and wouldnt mind learning why **not** to do this.

If you have a package that is meant to work across Dashing, Eloquent, Foxy, and Rolling, then having the backwards compatibility code is perfectly reasonable.  It will be executed on at least one of them.

---
@SebastianGrans CI is yellow because flake8 doesn't seem to like single-variable variable names.  Can you take a look and clean it up?

wjwwood commented on sensor_msgs_py/sensor_msgs_py/point_cloud2.py:
```diff
@@ -0,0 +1,229 @@
+                coordinates. [default: empty list]
+    @type  uvs: iterable
+    @return: Generator which yields a list of values for each point.
+    @rtype:  generator
```
nitpick: we use a different style when documenting parameters, and return type, etc... For example:

https://github.com/ros2/rclpy/blob/d7b4688743a99063e444f216f30c462fb03369b3/rclpy/rclpy/node.py#L126-L146

SebastianGrans replied:
Shall I change it while I'm at it?

clalancette replied:
Yeah, if you don't mind changing it, that would be great.  Thanks.

wjwwood reviewed (approved):
looks reasonable to me @clalancette, with one nitpick that shouldn't block it

wjwwood commented:
And with the linter failures addressed of course.

SebastianGrans commented:
@clalancette I am unable to replicate the flake8 error.

I did a fresh clone of my repo and ran build the exact build and test commands that are in the CI log [here](https://ci.ros2.org/job/ci_linux/13226/consoleText). Any recommendations? 

```bash 
grans@workhorse  ~/ws/src  rm -rf common_interfaces 
 grans@workhorse  ~/ws/src  git clone https://github.com/SebastianGrans/common_interfaces.git
Cloning into 'common_interfaces'...
remote: Enumerating objects: 23, done.
remote: Counting objects: 100% (23/23), done.
remote: Compressing objects: 100% (14/14), done.
remote: Total 2100 (delta 9), reused 22 (delta 9), pack-reused 2077
Receiving objects: 100% (2100/2100), 326.81 KiB | 405.00 KiB/s, done.
Resolving deltas: 100% (1395/1395), done.
 grans@workhorse  ~/ws/src  cd ..
 grans@workhorse  ~/ws  colcon build --base-paths "src" --build-base "build" --install-base "install" --event-handlers console_cohesion+ console_package_list+ --cmake-args -DBUILD_TESTING=ON --no-warn-unused-cli -DINSTALL_EXAMPLES=OFF -DSECURITY=ON --packages-up-to sensor_msgs_py
Topological order    
- std_msgs (ros.ament_cmake)
- geometry_msgs (ros.ament_cmake)
- sensor_msgs (ros.ament_cmake)
- sensor_msgs_py (ros.ament_python)
Starting >>> std_msgs
--- output: std_msgs                              
[omitted]
Finished <<< std_msgs [6.98s]
Starting >>> geometry_msgs
[omitted]
Finished <<< geometry_msgs [17.3s]
Starting >>> sensor_msgs
[omitted]
Finished <<< sensor_msgs [28.0s]
Starting >>> sensor_msgs_py
[omitted]
Finished <<< sensor_msgs_py [1.03s]

Summary: 4 packages finished [53.5s]
grans@workhorse  ~/ws  colcon test --base-paths "src" --build-base "build" --install-base "install" --event-handlers console_direct+ --executor sequential --retest-until-pass 2 --ctest-args -LE xfail --pytest-args -m "not xfail" --packages-select sensor_msgs_py
Starting >>> sensor_msgs_py
============================= test session starts ==============================
platform linux -- Python 3.8.5, pytest-6.0.1, py-1.9.0, pluggy-0.13.0
cachedir: /home/grans/ws/build/sensor_msgs_py/.pytest_cache
rootdir: /home/grans/ws/src/common_interfaces/sensor_msgs_py, configfile: pytest.ini
plugins: ament-copyright-0.9.5, launch-testing-ros-0.10.3, ament-pep257-0.9.5, ament-lint-0.9.5, launch-testing-0.10.3, ament-xmllint-0.9.5, ament-flake8-0.9.5, rerunfailures-9.0, repeat-0.8.0, colcon-core-0.6.1, cov-2.8.1
collecting ...                               
collected 9 items                                                              

test/test_copyright.py .                                                 [ 11%]
test/test_flake8.py .                                                    [ 22%]
test/test_pep257.py .                                                    [ 33%]
test/test_point_cloud2.py ......                                         [100%]

------ generated xml file: /home/grans/ws/build/sensor_msgs_py/pytest.xml ------
============================== 9 passed in 0.46s ===============================
Finished <<< sensor_msgs_py [1.22s]          

Summary: 1 package finished [1.44s]
 grans@workhorse  ~/ws  
```

clalancette commented:
> **EDIT:** I figured it out! It is the use of the small letter 'L' that is forbidden as described: [here](https://www.flake8rules.com/rules/E741.html). Still can't figure out why I it doesn't warn me when I build locally....

Yeah, I'm not sure why you couldn't reproduce locally.

> So it could be the use of the single letter iteration variable

Yes, I would guess that it is complaining about the single letter variable.  In point of fact, that rule that you link to explains exactly why; `l` is too easily confused with `1` or uppercase `I`, so it is better not to use it.  `f` isn't really easily confused with anything else, so it is not a problem.

---
In any case, I think once you fix the parameter comment style like @wjwwood suggests, and fix that flake8 warning, things are ready for another CI here.

flynneva commented:
tested locally @SebastianGrans's master branch and confirmed there was one warning but not due to the changes in this branch.

important to note the error shown in the CI wasnt visible. not sure what is going on there.

```
colcon test --base-paths "src" --build-base "build" --install-base "install" --event-handlers console_direct+ --executor sequential --retest-until-pass 2 --ctest-args -LE xfail --pytest-args -m "not xfail" --packages-select sensor_msgs_py
Starting >>> sensor_msgs_py
============================= test session starts ==============================
platform linux -- Python 3.8.5, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
cachedir: /home/flynn/code/ros/foxy/build/sensor_msgs_py/.pytest_cache
rootdir: /home/flynn/code/ros/foxy/src/common_interfaces/sensor_msgs_py, configfile: pytest.ini
plugins: ament-copyright-0.10.0, ament-pep257-0.10.0, ament-flake8-0.10.0, ament-lint-0.10.0, launch-testing-0.10.3, ament-xmllint-0.9.5, launch-testing-ros-0.10.3, rerunfailures-9.1.1, repeat-0.9.1, mock-1.10.4, cov-2.8.1, colcon-core-0.6.1
collecting ...                               
collected 9 items                                                              

test/test_copyright.py .                                                 [ 11%]
test/test_flake8.py .                                                    [ 22%]
test/test_pep257.py .                                                    [ 33%]
test/test_point_cloud2.py ......                                         [100%]

=============================== warnings summary ===============================
../../../../../../../../usr/lib/python3/dist-packages/pydocstyle/config.py:6
  Warning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.9 it will stop working

-- Docs: https://docs.pytest.org/en/stable/warnings.html
- generated xml file: /home/flynn/code/ros/foxy/build/sensor_msgs_py/pytest.xml -
========================= 9 passed, 1 warning in 0.94s =========================
--- stderr: sensor_msgs_py                   

=============================== warnings summary ===============================
../../../../../../../../usr/lib/python3/dist-packages/pydocstyle/config.py:6
  Warning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.9 it will stop working

-- Docs: https://docs.pytest.org/en/stable/warnings.html
---
Finished <<< sensor_msgs_py [1.92s]

Summary: 1 package finished [2.15s]
  1 package had stderr output: sensor_msgs_py
```

flynneva commented:
@clalancette is there a way to check what source code the CI is checking out? I couldnt find it anywhere on the jenkins dashboard.

wjwwood commented:
@flynneva it's in the console output, if you look there. Looking at the configuration of the job, you can find this file @clalancette is using too:

https://gist.githubusercontent.com/clalancette/0228a9b16b5fff7880a2405536971e08/raw/c222a10c98333531e6dfb951a79c82de2fceb27e/ros2.repos

It contains:

```
  ros2/common_interfaces:
    type: git
    url: https://github.com/SebastianGrans/common_interfaces.git
    version: master
```

So it's using master on your fork. If you want the commit hash being used, look in the console output of any of the jobs.

clalancette commented:
One of the issues with flake8 is that it only runs the checks that are currently installed.  So if you are missing one of the plugins that we use in CI, it will just silently skip the check.  That's my guess on what is going on here.

If you look at https://ci.ros2.org/job/ci_linux/13226/consoleFull , and search for `flake8`, you'll find the following list of python pip packages (including the flake8 plugins) that we are installing:

```
python3 -m pip install -U --force-reinstall EmPy coverage catkin_pkg flake8 flake8-blind-except flake8-builtins flake8-class-newline flake8-comprehensions flake8-deprecated flake8-docstrings flake8-import-order flake8-quotes importlib-metadata lark-parser mock mypy nose pep8 pydocstyle pyflakes pyparsing pytest pytest-cov pytest-mock pytest-repeat pytest-rerunfailures pytest-runner pyyaml vcstool colcon-core colcon-defaults colcon-library-path colcon-metadata colcon-mixin colcon-output colcon-package-information colcon-package-selection colcon-parallel-executor colcon-powershell colcon-python-setup-py colcon-recursive-crawl colcon-test-result colcon-cmake colcon-ros colcon-bash colcon-zsh
```

One of those is probably the problem.

flynneva commented:
@clalancette that was it! thank you so much!

@SebastianGrans, can confirm just changing that `l` variable to a `p` like the snippit below removes that flake8 error.

```
    return [Point._make(p) for p in read_points(cloud, field_names,
                                                skip_nans, uvs)]
```

clalancette commented:
Looking good to me.  Here's another shot at CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=13243)](http://ci.ros2.org/job/ci_linux/13243/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=8174)](http://ci.ros2.org/job/ci_linux-aarch64/8174/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10958)](http://ci.ros2.org/job/ci_osx/10958/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=13276)](http://ci.ros2.org/job/ci_windows/13276/)

clalancette commented:
All right, CI is all green here.  Thanks for the contribution, and for all of the iterations.  I'm going to merge this now.

dheera commented:
@SebastianGrans I was thinking that maybe it would be nice to have a numpy decoder for PointCloud2 as part of ROS to make things more efficient. I wrote one in ROSboard that reads the PointCloud2 entirely in a vectorized fashion with no loops over the points. See `def compressed_point_cloud2` in:

https://github.com/dheera/rosboard/blob/dev/rosboard/message_helper.py

Would you mind if I submit a pull request, and how should we make sure we don't break compatibility? Could it be a separate function e.g. `read_points_numpy`?

flynneva commented:
> Would we consider replacing read_points since ROS2 seems to embrace numpy all its subscribers, or would we want a separate function e.g. read_points_numpy?

if the inputs are the same to the new numpy 'read_points' version and the 'numpy' dependency can be easily be handled then I don't see why not to replace the logic with a faster alternative. We just would have to be careful to not affect end-users.

dheera commented:
> 'numpy' dependency can be easily be handled

I believe rosidl_generated_py itself depends on numpy already. I've gotten ros2 message data sometimes as array.array and sometimes as np.ndarray (I can't quite tell why rosidl picks one or the other instead of just using one all the time) but pretty sure numpy is a already dependency of ROS2 due to this.

In terms of not breaking logic one issue could be that trying to return a list of namedtuples would be counterproductive to numpy-ifying this, because that causes a memory reallocation and a Python list comprehension which is slow -- as opposed to directly returning the numpy array of structs. I believe the only difference is when returning the numpy array the end user would have to do point[0]['x'], point[0]['y'], etc. instead of point[0].x, point[0].y, etc.

But at the end of the day if this is a new library, is it a hard requirement to have the same API as ROS1? Considering ROS2 is already returning numpy arrays in messages one would think that keeping everything numpy and avoiding ever converting to a Python list would best embrace ROS2's advantages.

dheera commented:
Okay you know what, I figured out how to make it work. This is what is needed:
```
>>> array_of_my_structs = array_of_bytes.view(dtype = my_struct_type).view(dtype = np.recarray)
>>> array_of_my_structs[0].x
513
>>> array_of_my_structs[0].y
1027
```
So I think backwards compatibility with namedtuple users is possible! I'll work on a PR for this that supports skip_nans, uv etc., I think those should be easy to include.

dheera commented:
PR created here
https://github.com/ros2/common_interfaces/pull/155

I couldn't get the DCO working. Not sure what that is. I tried to follow the directions and I get
```
:) git rebase HEAD~3 --signoff
error: cannot rebase: Your index contains uncommitted changes.
error: Please commit or stash them.
```

I tried to follow the directions in the referenced link (https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally) and I get this
```
:) git fetch origin pull/155/head:master
fatal: couldn't find remote ref pull/155/head
```

an99990 commented:
so how do we extract points from PointCloud2 msg in ros2 ?  thank you

clalancette commented:
Please open questions like this on https://answers.ros.org, which is our central Question and Answer site.  You'll get a better answer there, and it will be searchable for the future.

Make sure to include a lot of information on what platform you are using, which ROS distribution you are using, and the exact steps you took.

SebastianGrans commented:
I'll answer this, but please tell me if I should avoid answering questions in old PRs. 

@an99990: I made a small demo a while ago on how to use this small utility package. Check it out here: https://github.com/SebastianGrans/ROS2-Point-Cloud-Demof

an99990 commented:
thank you @SebastianGrans :)
